ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,248 @@
+from ravenml.utils.io_utils import download_data_from_s3, copy_data_locally
+
+
+class DatasetWriter(object):
+
+    def __init__(self, create: CreateInput, **kwargs):
+        self.associated_files = kwargs['associated_files']
+        self.related_data_prefixes = kwargs['related_data_prefixes']
+        
+        metadata = create.metadata
+        self.num_folds = metadata['kfolds'] if metadata.get('kfolds') else 5 # Never actually used
+        self.test_percent = metadata['test_percent'] if metadata.get('test_percent') else .2
+        self.label_to_int_dict = {}
+        self.image_ids = None
+        self.filter_metadata = None
+        self.temp_dir = create.plugin_metadata['temp_dir_path']
+        self.dataset_path = create.dataset_path
+        self.dataset_name = metadata['dataset_name']
+        self.created_by = metadata['created_by']
+        self.comments = metadata['comments']
+        self.plugin_name = create.plugin_metadata['architecture']
+        self.imagesets_used = metadata['imagesets_used']
+        self.filter = metadata['filter']
+        self.cache = create.plugin_cache
+    
+    def construct_all(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def export_as_TFExample(self, *args, **kwargs):
+        raise NotImplementedError
+    
+    def write_additional_files(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def filter_and_load(self, metadata_prefix):
+        bucketConfig = get_config()
+        image_bucket_name = bucketConfig.get('image_bucket_name')
+
+        cli_spinner("Loading metadata...", ingest_metadata, self.imagesets_used, image_bucket_name, self.cache)
+
+        tags_df = load_metadata(metadata_prefix, self.cache)
+        filter_metadata = {"groups": []}
+
+        if self.filter:
+            image_ids = filter_sets()
+        else: 
+            image_ids = tags_df.index.tolist()
+                
+        # condition function for S3 download and local copying
+        def need_file(filename):
+            if len(filename) > 0:
+                image_id = filename[filename.index('_')+1:filename.index('.')]
+                if image_id in image_ids:
+                    return True
+
+        # if(data_source == "Local"):
+        #     cli_spinner("Copying data locally...", copy_data_locally, source_dir=kwargs["data_filepath"], 
+        #                 condition_func=need_file)
+        # elif(data_source == "S3"):
+
+        cli_spinner("Downloading data from S3...", download_data_from_s3, bucket_name=image_bucket_name, 
+                    filter_vals=self.imagesets_used, condition_func=need_file, cache=self.cache)
+
+        # sequester data for this specific run    
+        cache = self.cache
+        cache.ensure_clean_subpath('data/temp')
+        cache.ensure_subpath_exists('data/temp')
+
+        data_dir = cache.path / 'data'
+
+        cli_spinner("Copying data into temp folder...", copy_data_locally,
```

Comment:
Is there any way to avoid copying into a temp folder? I know this is the simplest way to do it for now. but it is a huge waste of resources if the imageset is large. It seems we should be able to just grab select images from the download folder directly.

Also raises a more important question: does the current implementation avoid downloading data again if not necessary? i.e, could a user have a bunch of imagesets cached, and make datasets from them quickly without redownload occuring?

Reply:
1. It would be possible to not copy into a temp folder. This would involve sending the plugin a list of paths directly to the files that it needs to deal with instead of a path to a temp folder with all the copied data it needs. The upside of this is it saves resources as you mentioned. However, if we send the plugin a list of paths into the cached imagesets, nothing is preventing the plugin from modifying the directory/files which could cause an issue down the line. 

2. The current implementation does cache imagesets with AWS sync. There should not be any redownloads occurring.

Reply:
to (1): I think it's ok to just send the plugin the data directly, and explicitly document in the "how to make a plugin stuff" that they should not alter the data directly. if a dataset is .5-5GB, copying it is so wasteful I think it's worth this risk